### DISCUSSION TITLE
What is the canonical way to determine the path to brew-managed binaries in non-interactive scripts?

### DISCUSSION THREAD

**Discussion by ratkins** [Homebrew/discussions#447] [Everyday usage]
I'm writing some scripts which need to be run on both M1 and Intel-based macOS machines.

Since non-interactive scripts should explicitly specify paths to binaries, the script needs to work out if the current machine's brew installation is in `/opt/homebrew` on an M1, or `/usr/local/bin` on Intel.

What is the canonical way of doing this? Is it as simple as something like:
```
if [ -d "/opt/homebrew/bin" ]
then
    HOMEBREW_BIN="/opt/homebrew/bin"
else
    HOMEBREW_BIN="/usr/local/homebrew/bin"
fi

"$HOMEBREW_BIN/foo" --arg1 bar
#... etc
```

That's still pretty tedious to write at the top of every script, or separate out into a function in a file that has to be sourced in every script.

_(Imagine for the purposes of this discussion that in the near-enough future, most people on M1 machines will only have one Homebrew installation, or at least if they have both, scripts should prefer the `/opt/homebrew` installation.)_

**Comment by gromgit** [comment]
Just use `brew shellenv` to set up the environment for you (including `$PATH`), then you don't have to qualify anything:
```
# Favor native ARM over Rosetta'd Intel on M1
for d in /opt/homebrew /usr/local; do
  if [[ -x ${d}/bin/brew ]]; then
    eval "$(${d}/bin/brew shellenv)"
    break
  fi
done
```

**Reply by ratkins** [reply]
Ah, ok, I guess that's better but it's still pretty wordy.

**Reply by gromgit** [reply]
Dunno about you, but for stuff I want to be done in every script, I put it in a library script, then everything else gets a single line up top:
```
[[ -r ~/lib/my.sh ]] && . ~/lib/my.sh
```
Easy peasy.
